src/Interpreters/TreeRewriter.cpp

```diff
@@ -325,6 +326,39 @@ struct ExistsExpressionData
 
 using ExistsExpressionVisitor = InDepthNodeVisitor<OneTypeMatcher<ExistsExpressionData>, false>;
 
+struct ReplacePositionalArgumentsData
+{
+    using TypeToVisit = ASTSelectQuery;
+    ContextPtr context;
+
+    void visit(ASTSelectQuery & select_query, ASTPtr &) const
+    {
+        if (context->getSettingsRef().enable_positional_arguments)
```

Comment:
We can check the setting in `TreeRewrtier::normalize` and do not apply `ReplacePositionalArgumentsVisitor` at all if `enable_positional_arguments = false` to improve readability a bit.